### DISCUSSION TITLE
Parallel instructions

### DISCUSSION THREAD

**Discussion by mumbel** [NationalSecurityAgency/ghidra#6084] [Q&A]
I don't know if I've come across this in another processor.  SuperH DSP extension has parallel instructions, basically two "flavors" (and one can precede the other).

The first changes the decoding of the following subset of DSP instructions based on preceding subset of DSP instructions.  The 2nd instruction conflicts with non DSP instructions.  A context can wrap around these set and everything else unset.  I think this makes sense.

The second run X and Y memory operations in the same instruction and decodes two operations from the 16-bits.
- either a `nop` or some variant of around seven `movx.w` for X registers
- either a `nop` or some variant of around seven `movy.w` for Y registers
- mov registers have things like deref and post increment and flip src/dst 

examples:
-  `nopx` and `nopy`
- `nopx` and `movy.w Ay,Dy`
- `movx.w Ax,Dx` and `movy.w Ay,Dy`

My current solution is eight xymemop and eight ymemop constructions and two instructions
```
:xmemop/ymemop is xmemop & ymemop & opcode_11=0 {}
:xmemop/ymemop is xmemop & ymemop & opcode_11=1 [parallel = 1; globalset(inst_next, parallel); ]
```

This results in one mnemonic and no operands.  For PCODE this is fine, for listing this result isnt great, but usable (this many mnemonics an issue?  permutations of:  8 X mnemonics, 8 Y mnemonics, num Ax regs, num Ay regs, num Dx regs, num Dy regs).

Is there any way to get `movx.w x0,a0/movy.w y0,m0` using a 2nd line or split them out as operands so the mnemonic string is not so long using tables like this, otherwise is that just 64 instructions (and then double for the parallel... i was unable to `... & opcode_11 [ parallel=opcode_11; ... ]` ... is the right or was I doing something weird)?

maybe 4 sets of tables,  operationX (mnemonic, sleigh), operationY (mnemonic, sleigh), reg display X (operands, no sleigh), reg display Y (operands, no sleigh) and then instruction is `:xmemop/ymemop xmemreg/ymemreg`  (... with some combinations of `^`?) to get 64 mnemonics but they get operands?

But wanting and don't think its possible:   

```
00001234  AB CD  movx.w x0 , a0
                 movy.w y0 , m0 
```

**Comment by GhidorahRex** [comment]
What you most likely want to do is use `crossbuild`. Unfortunately, it's poorly documented, I don't really understand it well, and I have no examples I can find to show you. However I'm happy to work with you on it, and maybe we can use the SH3-DSP as an example for future reference on the feature.

`crossbuild` takes an address and a pcode label. e.g. `crossbuild CrossBuildAddress, COMMIT;` The address tells the crossbuild where to pull the bytes for the parallel instruction from, and the label tells crossbuild where to slot the pcodeops from the parallel instruction into the first. Looking at the SH3-DSP instructions, I don't think it should be terribly complicated once we get it working. You might need to play around with +/- a byte or two to find the right spot to pull bytes in.

Here's a dummy example I cooked up that may or may not work properly:
```
CrossBuildAddress: loc is format=x; [loc = inst_start-2;] { local tmp:4 = loc; export *[ram]:4 loc;} #inst_start -2? Not sure this is right here.
checkParallel: "" is phase=0 [phase=1;] {}
checkParallel: "" is phase=1 & format=x & CrosBuildAddress [phase=2;] { crossbuild CrossBuildAddress, COMMIT; }

:INST_MNEMONIC result, operand is bits=y & operand & result & checkParallel {
local tmp = operand;
build checkParallel
<<COMMIT>>
result = tmp;
}
```

With the SH-3 DSP you would want to check the instruction format, and if it matches the parallel instruction format include the crossbuild directive.

The way this will look in the listing is you will have:
```
movex.w x0, a0 ||
movey.w y0, m0
```

with the bars identifying that the instructions should occur in parallel. But I haven't been able to test any of that. And hopefully it gets you started in the right direction.

**Reply by mumbel** [reply]
Awesome, I had not come across crossbuild yet. I'll try to mess with it.

@GhidorahRex   this is for sh2.  Undecided how much more effort I'll put in, for now I think it's usable, but if I implement dsp and get it working well enough, interest in an MR combining super and superh4 modules along with that feature or just limit to maybe a superh PR?   Is 2 modules a feature not a bug at this point?

**Reply by mumbel** [reply]
https://gist.github.com/mumbel/2d37576d5870af25eff8984b4f5a95df

I got lost by the time I read `format=x` lol...  here is more to go off of if its easier with my current approach mixed with your suggestion.  I'm not 100% on my original implementation, working from memory from my work machine where I got it going.

**Reply by mumbel** [reply]
I spent a little time trying again tonight and I didn't have any luck getting this going.  Although it does looks like `inst_start` is correct here (vs `inst_start-2`).  DebugSleighInstruction didn't give any hints, don't know if it would catch this logic or not.

**Reply by GhidorahRex** [reply]
Can you show me a sample instruction byte segment for one of the parallel instructions?

**Reply by mumbel** [reply]
I'll try to get my working module online with dsp registers and context, but until then `nopx` and `nopy` is just `f0 00`

**Reply by mumbel** [reply]
@GhidorahRex   https://gist.github.com/mumbel/3305812e54e46ac7bbf97fc312610e0d

unless I have bugs, here is an example with expected instructions: `f8 00 95 37 f8 02 a1 a7 fa 28f f9 08 f8 00 17 a8`
```
f8 00    nopx/nopy
95 37    pand a0,m1,a0
f8 02    nopx/movy.w @r6+,y0
a1 a7    psub a0,m0,a0
fa 28    movx.w a0,@r5+/nopy
f9 08    movx.w @r4+,x0/movy.w @r7+,@y0
f8 00    nopx/nopy
17 a8    pshl #-0x6,x0
```

**Reply by GhidorahRex** [reply]
Based on my understanding of the manual, anything in the f8-fb range is a parallel processing instruction, so those are easy to weed out.

I would suggest stripping out everything but the double- and single-transfers and getting those to work properly, then introducing the parallel instructions once you get the `movx.w` and `movy.w` instructions working.


```
:^instruction is phase=0 & opcode_10_15=0b111110 & parallel=0 [parallel=1; globalset(inst_next, parallel); phase=1;] {build instruction;}
CrossBuildAddress: loc is phase=1 [loc = inst_start-2;] { local tmp:4 = loc; export *[ram]:4 loc;}
CrossBuildAddress: loc is phase=2 [loc = inst_start;] { local tmp:4 = loc; export *[ram]:4 loc;}

#checkParallel check for movy
checkParallel: "" is phase=1 & CrossBuildAddress [phase=2;] { crossbuild CrossBuildAddress, MOVY; }

#check parallel check for parallel instructions
checkParallel: "" is phase=2 & opcode_10_11=2 & CrossBuildAddress [phase=3;] { crossbuild CrossBuildAddress, PARA_INST; }
checkParallel: "" is phase=2 & [phase=3;] { } # double-transfer only, no parallel

#all the movx transfers...
:movx.w @ax,dx is opcode_12_15=0b1111 & opcode_10=0 & ax & dx & opcode_05=0 & opcode_02_03=0b01 & checkParallel & phase=1
[ phase = 2; globalset(inst_next, phase); ]
{
    dx = ax << 16;
    <<MOVY>>
}
#etc...

#all the movy transfers...

:movy.w @ay,dy is opcode_12_15=0b1111 & opcode_10=0 & ay & dy & opcode_04=0 & opcode_00_01=0b01 & checkParallel & phase=2
{
    dy = ay << 16;
    <<PARA_INST>>
}
#etc...

# All the parallel instructions:
:pabs sx,dz is parallel=1 & phase=3 & opcode_08_15=0b10001000 & sx & opcode_04_05=0b00 & dz
{
    local tmp:1 = (sx s>= 0);
    dz = (zext(tmp != 0) * sx) + (zext(tmp == 0) * -sx);
    #TODO  FLAGS
}
#etc...
```

and ideally, this would show up in the listing as something like:

```
f8 00    nopx ||
         nopy ||
95 37    pand a0,m1,a0
f8 02    nopx ||
         movy.w @r6+,y0 ||
a1 a7    psub a0,m0,a0
fa 28    movx.w a0,@r5+ ||
         nopy
f9 08    movx.w @r4+,x0 ||
         movy.w @r7+,@y0
f8 00    nopx ||
         nopy
17 a8    pshl #-0x6,x0 ||
```

**Reply by mumbel** [reply]
Cool, I'll take another try this weekend, thanks for the tips

**Reply by mumbel** [reply]
Tried again for a bit and no luck.  Pushed changes to https://github.com/mumbel/ghidra/tree/superh_dsp

I think your suggestions mostly made sense, but I did have some tweaks. 
 DebugSleighInstruction seemed to at least be getting the details right for the first instruction, but that script cant follow the crossbuild.

I'm not even able to find the java that inserts the `||` in the listing, the only function that sort of looked worth tracing at a glance (`hasCrossBuildDependency`) had no callers

**Reply by GhidorahRex** [reply]
I'm seeing the same thing. Crossbuild may not work for this then. I don't know of a way to split the data moves into separate lines otherwise. But the parallel instruction part looks good for now. I'll keep playing around with it.
